group-coordinator/src/main/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignor.java

```diff
@@ -0,0 +1,527 @@
+        final String processId = memberIdToProcessId.get(memberId);
+
+        // add the tasks to the corresponding ProcessSpec
+        processIdToProcessSpec.get(processId).addTask(memberId, subtopologyId, taskId);
+        // add all pair combinations: update taskPairs
+        addToTaskPairs(memberId, subtopologyId, taskId);
+    }
+
+    private void maybeRemoveExtraTasks(Map<String, Set<Integer>> tasks) {
+        int activeTasksCount = tasks.values().stream().mapToInt(Set::size).sum();
+        if (activeTasksCount > tasksPerMember) {
+            int curActiveTasksCount = 0;
+            for (Map.Entry<String, Set<Integer>> entry : tasks.entrySet()) {
+                int remaining = tasksPerMember - curActiveTasksCount;
+                if (curActiveTasksCount < tasksPerMember) {
+                    entry.setValue(entry.getValue().stream()
+                            .skip(0) // start offset
+                            .limit(Math.min(remaining, entry.getValue().size()))
+                            .collect(Collectors.toSet()));
+                    curActiveTasksCount += entry.getValue().size();
+                } else { // remove the extra tasks
+                    entry.setValue(new HashSet<>());
+                }
+            }
+        }
+    }
+
+    private void addToTaskPairs(String memberId, String subtopologyId, int taskId) {
+        final String processId = memberIdToProcessId.get(memberId);
+        taskPairs.addPairs(new TaskId(subtopologyId, taskId), processIdToProcessSpec.get(processId).assignedTasks());
+    }
+
+    private void addToTaskPairs(Map<String, Set<Integer>> curActiveTasks) {
+        List<TaskId> taskList = new ArrayList<>();
+        for (Map.Entry<String, Set<Integer>> entry : curActiveTasks.entrySet()) {
+            String subtopologyId = entry.getKey();
+            for (Integer id : entry.getValue()) {
+                taskList.add(new TaskId(subtopologyId, id));
+            }
+        }
+
+        for (int i = 1; i < taskList.size(); i++) {
+            taskPairs.addPairs(taskList.get(i - 1), new HashSet<>(taskList.subList(i, taskList.size())));
+        }
+    }
+
+    private int computeTasksPerMember(int allTasks, int totalCapacity) {
+        if (totalCapacity == 0) {
+            return 0;
+        }
+        int tasksPerMember = allTasks / totalCapacity;
+        if (allTasks - (tasksPerMember * totalCapacity) > 0) {
+            tasksPerMember++;
+        }
+        return tasksPerMember;
+    }
+
+    private String findMemberWithLeastLoad(Set<String> members, String subtopologyId, int taskId, boolean isLimited) {
+        Set<String> processes = new HashSet<>();
+        for (String member: members) {
+            processes.add(memberIdToProcessId.get(member));
+        }
+        // find the set of right pairs
+        Set<String> rightPairs = findRightPairs (processes, new TaskId(subtopologyId, taskId));
```

Comment:
Could you explain the logic behind "findRIghtPairs"? I don't think I really understand this.